### PR TITLE
Columnar: refactor read path and fix zero-column tables.

### PR DESCRIPTION
**Pull request by jeff-davis**

### CODE REVIEW

**Review by onurctirtir — COMMENTED** [review]
Today I just read this file to learn more about columnar.
Then I thought that reviewing this pr would be a good opportunity for me to learn the things :), left some minor comments & some questions:

File: src/backend/columnar/cstore_reader.c [hunk 1]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+ * columnValues and return true. Otherwise, return false.
+ */
+static bool
+ReadChunkGroupNextRow(ChunkGroupReadState *chunkGroupReadState, Datum *columnValues,
+					  bool *columnNulls)
 {
-	MemoryContextDelete(readState->stripeReadContext);
-	list_free_deep(readState->stripeList);
-	pfree(readState);
+	if (chunkGroupReadState->currentRow >= chunkGroupReadState->rowCount)
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:453]: No. I could change the ``>`` into an Assert instead, but then it'll be compiled out.

File: src/backend/columnar/cstore_reader.c [hunk 2]
```diff
@@ -142,16 +175,10 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
 	readState->stripeList = stripeList;
 	readState->projectedColumnList = projectedColumnList;
 	readState->whereClauseList = whereClauseList;
-	readState->stripeBuffers = NULL;
-	readState->readStripeCount = 0;
-	readState->stripeReadRowCount = 0;
 	readState->chunkGroupsFiltered = 0;
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:178]: Isn't the number of chunk groups filtered a property of the entire scan? What does being an integer have to do with it?

[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/columnar/cstore_reader.c:178]: oh true, looks like I misunderstood the reason

File: src/backend/columnar/cstore_reader.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/columnar/cstore_reader.c:453]: then I guess keeping that as is would be safer

[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:453]: I added the Assert inside the branch to be safer.

File: src/backend/columnar/cstore_reader.c [hunk 3]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+
+/*
+ * BeginStripeRead
+ *
+ * Allocate state for reading a stripe.
+ */
+static StripeReadState *
+BeginStripeRead(StripeMetadata *stripeMetadata, Relation rel, TupleDesc tupleDesc,
+				List *projectedColumnList, List *whereClauseList, MemoryContext
+				stripeContext)
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:267]: Done.

File: src/backend/columnar/cstore_reader.c [hunk 4]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+																   whereClauseList,
+																   &stripeReadState->
+																   chunkGroupsFiltered);
+
+		stripeReadState->rowCount = stripeReadState->stripeBuffers->rowCount;
+	}
+	else
+	{
+		stripeReadState->stripeBuffers = NULL;
+
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:306]: Is that covered by the regression tests added here: 52297804ae08508a106dce125aa0b2b9276f8b09 ? What other tests would you suggest?

File: src/backend/columnar/cstore_reader.c [hunk 5]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+EndStripeRead(StripeReadState *stripeReadState)
+{
+	pfree(stripeReadState);
+}
 
-		/* if we have read all stripes, return false */
-		if (readState->readStripeCount == stripeCount)
+
+/*
+ * ReadStripeNextRow
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:332]: DOne.

File: src/backend/columnar/cstore_reader.c [hunk 6]
```diff
@@ -63,25 +89,32 @@ struct TableReadState
...
+										 List *whereClauseList, MemoryContext
+										 stripeContext);
+static void EndStripeRead(StripeReadState *stripeReadState);
+static bool ReadStripeNextRow(StripeReadState *stripeReadState, Datum *columnValues,
+							  bool *columnNulls);
+static ChunkGroupReadState * BeginChunkGroupRead(StripeBuffers *stripeBuffers, int
+												 chunkIndex,
+												 TupleDesc tupleDesc,
+												 List *projectedColumnList,
+												 MemoryContext cxt);
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:107]: Done.

File: src/backend/columnar/cstore_reader.c [hunk 7]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+	while (true)
+	{
+		if (stripeReadState->chunkGroupReadState == NULL)
 		{
-			readState->stripeBuffers = stripeBuffers;
-			readState->stripeReadRowCount = 0;
-			readState->deserializedChunkIndex = -1;
-			break;
+			stripeReadState->chunkGroupReadState = BeginChunkGroupRead(
+				stripeReadState->stripeBuffers,
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:371]: Something to consider, though passing entire structs can obscure what data is actually needed by the callee and what information should be hidden from the callee.

File: src/backend/columnar/cstore_reader.c [hunk 8]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+		return false;
+	}
+
+	/*
+	 * Initialize to all-NULL. Only non-NULL projected attributes will be set.
+	 */
+	memset(columnNulls, true, sizeof(bool) * chunkGroupReadState->columnCount);
+
+	ListCell *projectedColumnCell;
+	foreach(projectedColumnCell, chunkGroupReadState->projectedColumnList)
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:464]: Done.

File: src/backend/columnar/cstore_reader.c [hunk 9]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+ * sets non-NULL entries.
+ *
+ * If more rows can be read from the current stripe, fill in non-NULL
+ * columnValues and return true. Otherwise, return false.
+ */
+static bool
+ReadStripeNextRow(StripeReadState *stripeReadState, Datum *columnValues,
+				  bool *columnNulls)
+{
+	if (stripeReadState->currentRow >= stripeReadState->rowCount)
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:344]: Added Assert.

File: src/backend/columnar/cstore_reader.c [hunk 10]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
 		}
-	}
 
-	uint32 chunkIndex = readState->stripeReadRowCount / stripeMetadata->chunkRowCount;
-	uint32 chunkRowIndex = readState->stripeReadRowCount % stripeMetadata->chunkRowCount;
+		if (!ReadChunkGroupNextRow(stripeReadState->chunkGroupReadState, columnValues,
+								   columnNulls))
+		{
+			EndChunkGroupRead(stripeReadState->chunkGroupReadState);
+			stripeReadState->chunkGroupReadState = NULL;
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:386]: Done.

File: src/backend/columnar/cstore_reader.c [hunk 11]
```diff
@@ -165,134 +192,291 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+}
+
+
+/*
+ * EndStripeRead
+ */
+static void
+EndStripeRead(StripeReadState *stripeReadState)
+{
+	pfree(stripeReadState);
```
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/cstore_reader.c:327]: ``projectedColumnList`` is a borrowed reference, so I can't free it there. Potentially at the end of a scan, but it's allocated in the scan context anyway.

``stripeBuffers`` is allocated in the ``stripeReadContext``, so will be freed when that context is reset. 

At least, that's how things are supposed to work. If something seems off and you see a leak, let me know.

[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/columnar/cstore_reader.c:327]: thanks for the explanation

File: src/backend/columnar/cstore_reader.c [hunk 12]
```diff
@@ -165,134 +192,289 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+void
+ColumnarRescan(TableReadState *readState)
+{
+	readState->stripeReadState = NULL;
+	readState->currentStripe = 0;
+	readState->chunkGroupsFiltered = 0;
+}
+
+
+/* Finishes a columnar read operation. */
```
Comment (onurctirtir) [inline src/backend/columnar/cstore_reader.c:249]: similar to https://github.com/citusdata/citus/pull/4668#discussion_r580366632

File: src/backend/columnar/cstore_reader.c [hunk 13]
```diff
@@ -165,134 +192,289 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+ColumnarEndRead(TableReadState *readState)
+{
+	MemoryContextDelete(readState->stripeReadContext);
+	list_free_deep(readState->stripeList);
+	pfree(readState);
+}
+
+
+/*
+ * BeginStripeRead
```
Comment (onurctirtir) [inline src/backend/columnar/cstore_reader.c:260]: similar to https://github.com/citusdata/citus/pull/4668#discussion_r580366632

File: src/backend/columnar/cstore_reader.c [hunk 14]
```diff
@@ -165,134 +192,289 @@ ColumnarBeginRead(Relation relation, TupleDesc tupleDescriptor,
...
+ *
+ * Allocate state for reading a stripe.
```
Comment (onurctirtir) [inline src/backend/columnar/cstore_reader.c:262]: similar to https://github.com/citusdata/citus/pull/4668#discussion_r580366632

**Review by onurctirtir — APPROVED** [review]
Left two minor comments regarding function comments

File: src/backend/columnar/cstore_reader.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/columnar/cstore_reader.c:306]: >Is that covered by the regression tests added here: 5229780 ?

oh true, we already have such tests that are added in that commit

**Pull request merged**